### DISCUSSION TITLE
Upgrading to React 17

### DISCUSSION THREAD

**Discussion by eatyourgreens** [zooniverse/front-end-monorepo#2301] [General]
[Next 11](https://nextjs.org/blog/next-11) came out in June, and [requires React 17 as a minimum](https://nextjs.org/docs/messages/react-version). I've opened #2281 for the upgrade.

[React 17](https://reactjs.org/blog/2020/10/20/react-v17.html) was released in October 2020 and adds no new features to React. It's a stepping stone, intended to make it easier to gradually upgrade to future versions of React. [React 18](https://reactjs.org/blog/2021/06/08/the-plan-for-react-18.html) was announced a couple of weeks ago, and does include changes that we're probably going to need for SSR. 18 is out in alpha now, so that library authors can work on upgrades.

We use [enzyme](https://enzymejs.github.io/enzyme/) to test React components. An enzyme adapter for React 17 was started in August 2020 (when React 17 was first released) but is [still in development](https://github.com/enzymejs/enzyme/pull/2430). The enzyme adapter is the work of a single person, and they have publicly asked for [help to complete it](https://github.com/enzymejs/enzyme/pull/2430#issuecomment-827817232). In the interim, [an unofficial adapter](https://github.com/wojtekmaj/enzyme-adapter-react-17) is available. The developer of the unofficial adapter has said that their adapter will only be supported until the official adapter is available. They are also working on the PR to release an official adapter.

I've used the unofficial React 17 adapter in #2281, but opened this discussion to talk about whether we continue to use enzyme, or move to another testing library eg. [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

**Comment by eatyourgreens** [comment]
`visx` are having a similar discussion: https://github.com/airbnb/visx/issues/872

**Comment by srallen** [comment]
Can the focus of this discussion be specifically on what it would take to migrate away from Enzyme and what the alternatives are? The issue isn't upgrading React 17. We know we need and want to do that. The issue is the waning support of Enzyme and the impact that is having on test writing and support.

**Reply by eatyourgreens** [reply]
It can. That's why I've linked to the `visx` discussion of whether they're going to stick with enzyme. (TLDR: they'll move to RTL for tests that are broken in enzyme.)

**Reply by srallen** [reply]
This seems more like a stop gap rather than a resolution to me. Something that could be done in a migration to something else as it doesn't really address the maintenance concerns of the project.

**Reply by eatyourgreens** [reply]
Well, they do need to ship support for React 17 and enzyme is blocking that.
https://github.com/airbnb/visx/issues/872#issuecomment-871675714

**Comment by srallen** [comment]
Ok, so let's be focused on what we want to do. 

The options as it seems to me is going to be:

- chai for just doing assertions on functions (already doing)
- react test utils for unit testing on react components as needed (which is already available to us with `react-dom`)
- react-testing-library for rendered output and behavior testing on react components

In the past a strategy we've tried to do is write new things with the new library or tool and swap over old things when we need to make edits, but I think that usually results in a hybrid that never gets finished. What I'm going to suggest since many (all?) of us are wrapping up development cycles on certain feature sets right now is to have a focus week on just doing a conversion from enzyme to react-testing-library, see how far we get, and then evaluate how to go from there: maybe schedule another maintenance focus week with the whole team in the near future or swap to the gradual approach.

**Comment by shaunanoordin** [comment]
Since #2281 has been merged & deployed due to an urgent security fix, how do we archive this discussion, or set it as resolved? (I see "lock conversation" and "delete discussion" in the menu.)

For those just coming in to this thread, since the last on 14 Jul...

- We've tested and added `react-testing-library` to FEM
- A lot of existing FEM components still use Enzyme, but Enzyme is going to be gradually phased out in favour of RTL
  - There's also some progress on the Enzyme adapter recently, so that's good too?
- We've upgraded to React 17 _and_ Next.js 11.1

**Reply by srallen** [reply]
I think the discussion title is a misnomer. This discussion really is about enzyme to RTL migration, so with that scope, I'd like to keep it "open" so to speak because I think the adapter issue is still unresolved and I hope to do another maintenance week in the coming weeks where we might get a sense for pacing of the migration. The initial week we did focusing on this was mostly about learning, so not a good measure for work pace.

**Comment by eatyourgreens** [comment]
Visx officially supports React 17 now, so we should probably upgrade. https://github.com/airbnb/visx/pull/1268

**Comment by eatyourgreens** [comment]
React 18 is out of beta now, and the `createRoot` API can't be used with Enzyme.
https://dev.to/wojtekmaj/enzyme-is-dead-now-what-ekl

This gives a useful overview of the background and context.
https://www.piotrstaniow.pl/goodbye-enzyme

There's an interesting quote from Rachel Nabors, in that article. Maybe we should discourage class components in new code?
>Class components are going to be around for years to come—for example, there are tens of thousands in production at Facebook already. However, we do recommend that new apps be built with function components and Hooks, which is why we want those docs front and center. The class component docs will remain available for folks working with those components, and class components themselves might one day be spun out into their own package—but if that did happen, we're provide migration scripts to automate that transition :)
